File: backend/services/save_service.py
from sqlalchemy.orm import Session
from sqlalchemy import and_, or_, desc
from typing import Optional, List, Dict, Any, Tuple
from database.models import SavedCode, SavedLog, SavedSession, User
from models.save_schemas import SaveCodeRequest, SaveLogRequest, SaveSessionRequest
import uuid
from datetime import datetime


class SaveService:
    def __init__(self):
        pass
    
    # Code Save Operations
    def save_code(self, db: Session, user_id: str, request: SaveCodeRequest) -> SavedCode:
        """Save a code snippet to the database."""
        saved_code = SavedCode(
            id=str(uuid.uuid4()),
            title=request.title,
            description=request.description,
            code_content=request.code_content,
            language=request.language,
            tags=request.tags,
            user_id=user_id
        )
        
        db.add(saved_code)
        db.commit()
        db.refresh(saved_code)
        return saved_code
    
    def get_saved_codes(
        self, 
        db: Session, 
        user_id: str, 
        page: int = 1, 
        per_page: int = 20,
        search: Optional[str] = None,
        language: Optional[str] = None,
        tags: Optional[List[str]] = None
    ) -> Tuple[List[SavedCode], int]:
        """Get saved code snippets with pagination and filtering."""
        query = db.query(SavedCode).filter(SavedCode.user_id == user_id)
        
        # Apply filters
        if search:
            query = query.filter(
                or_(
                    SavedCode.title.ilike(f"%{search}%"),
                    SavedCode.description.ilike(f"%{search}%"),
                    SavedCode.code_content.ilike(f"%{search}%")
                )
            )
        
        if language:
            query = query.filter(SavedCode.language == language)
        
        if tags:
            # For JSON array contains search
            for tag in tags:
                query = query.filter(SavedCode.tags.contains([tag]))
        
        # Get total count
        total = query.count()
        
        # Apply pagination and ordering
        offset = (page - 1) * per_page
        codes = query.order_by(desc(SavedCode.created_at)).offset(offset).limit(per_page).all()
        
        return codes, total
    
    def get_saved_code(self, db: Session, user_id: str, code_id: str) -> Optional[SavedCode]:
        """Get a specific saved code snippet."""
        return db.query(SavedCode).filter(
            and_(SavedCode.id == code_id, SavedCode.user_id == user_id)
        ).first()
    
    def delete_saved_code(self, db: Session, user_id: str, code_id: str) -> bool:
        """Delete a saved code snippet."""
        code = self.get_saved_code(db, user_id, code_id)
        if code:
            db.delete(code)
            db.commit()
            return True
        return False
    
    # Log Save Operations
    def save_log(self, db: Session, user_id: str, request: SaveLogRequest) -> SavedLog:
        """Save logs to the database."""
        saved_log = SavedLog(
            id=str(uuid.uuid4()),
            title=request.title,
            description=request.description,
            log_content=request.log_content,
            log_level=request.log_level,
            source=request.source,
            log_metadata=request.log_metadata,
            tags=request.tags,
            user_id=user_id
        )
        
        db.add(saved_log)
        db.commit()
        db.refresh(saved_log)
        return saved_log
    
    def get_saved_logs(
        self, 
        db: Session, 
        user_id: str, 
        page: int = 1, 
        per_page: int = 20,
        search: Optional[str] = None,
        log_level: Optional[str] = None,
        source: Optional[str] = None,
        tags: Optional[List[str]] = None
    ) -> Tuple[List[SavedLog], int]:
        """Get saved logs with pagination and filtering."""
        query = db.query(SavedLog).filter(SavedLog.user_id == user_id)
        
        # Apply filters
        if search:
            query = query.filter(
                or_(
                    SavedLog.title.ilike(f"%{search}%"),
                    SavedLog.description.ilike(f"%{search}%"),
                    SavedLog.log_content.ilike(f"%{search}%")
                )
            )
        
        if log_level:
            query = query.filter(SavedLog.log_level == log_level)
        
        if source:
            query = query.filter(SavedLog.source.ilike(f"%{source}%"))
        
        if tags:
            # For JSON array contains search
            for tag in tags:
                query = query.filter(SavedLog.tags.contains([tag]))
        
        # Get total count
        total = query.count()
        
        # Apply pagination and ordering
        offset = (page - 1) * per_page
        logs = query.order_by(desc(SavedLog.created_at)).offset(offset).limit(per_page).all()
        
        return logs, total
    
    def get_saved_log(self, db: Session, user_id: str, log_id: str) -> Optional[SavedLog]:
        """Get a specific saved log."""
        return db.query(SavedLog).filter(
            and_(SavedLog.id == log_id, SavedLog.user_id == user_id)
        ).first()
    
    def delete_saved_log(self, db: Session, user_id: str, log_id: str) -> bool:
        """Delete a saved log."""
        log = self.get_saved_log(db, user_id, log_id)
        if log:
            db.delete(log)
            db.commit()
            return True
        return False
    
    # Session Save Operations
    def save_session(self, db: Session, user_id: str, request: SaveSessionRequest) -> SavedSession:
        """Save a complete debug session to the database."""
        saved_session = SavedSession(
            id=str(uuid.uuid4()),
            title=request.title,
            description=request.description,
            session_data=request.session_data,
            user_id=user_id
        )
        
        db.add(saved_session)
        db.commit()
        db.refresh(saved_session)
        return saved_session
    
    def get_saved_sessions(
        self, 
        db: Session, 
        user_id: str, 
        page: int = 1, 
        per_page: int = 20,
        search: Optional[str] = None
    ) -> Tuple[List[SavedSession], int]:
        """Get saved sessions with pagination and filtering."""
        query = db.query(SavedSession).filter(SavedSession.user_id == user_id)
        
        # Apply search filter
        if search:
            query = query.filter(
                or_(
                    SavedSession.title.ilike(f"%{search}%"),
                    SavedSession.description.ilike(f"%{search}%")
                )
            )
        
        # Get total count
        total = query.count()
        
        # Apply pagination and ordering
        offset = (page - 1) * per_page
        sessions = query.order_by(desc(SavedSession.created_at)).offset(offset).limit(per_page).all()
        
        return sessions, total
    
    def get_saved_session(self, db: Session, user_id: str, session_id: str) -> Optional[SavedSession]:
        """Get a specific saved session."""
        return db.query(SavedSession).filter(
            and_(SavedSession.id == session_id, SavedSession.user_id == user_id)
        ).first()
    
    def delete_saved_session(self, db: Session, user_id: str, session_id: str) -> bool:
        """Delete a saved session."""
        session = self.get_saved_session(db, user_id, session_id)
        if session:
            db.delete(session)
            db.commit()
            return True
        return False

File: frontend/src/services/saveService.ts
interface SaveCodeRequest {
  title: string;
  description?: string;
  code_content: string;
  language?: string;
  tags?: string[];
}

interface SaveLogRequest {
  title: string;
  description?: string;
  log_content: string;
  log_level?: string;
  source?: string;
  log_metadata?: Record<string, any>;
  tags?: string[];
}

interface SaveSessionRequest {
  title: string;
  description?: string;
  session_data: Record<string, any>;
}

interface SaveResponse {
  success: boolean;
  message: string;
  item_id?: string;
  item?: any;
}

interface SavedItem {
  id: string;
  title: string;
  description?: string;
  created_at: string;
  updated_at?: string;
}

interface SavedCode extends SavedItem {
  code_content: string;
  language?: string;
  tags?: string[];
}

interface SavedLog extends SavedItem {
  log_content: string;
  log_level?: string;
  source?: string;
  log_metadata?: Record<string, any>;
  tags?: string[];
}

interface SavedItemsResponse {
  items: SavedItem[];
  total: number;
  page: number;
  per_page: number;
  total_pages: number;
}

class SaveService {
  private baseUrl: string;

  constructor() {
    this.baseUrl = process.env.REACT_APP_BACKEND_URL || 'http://localhost:8000';
  }

  private getAuthHeaders(): Record<string, string> {
    const token = localStorage.getItem('auth_token');
    const headers: Record<string, string> = {
      'Content-Type': 'application/json',
    };
    
    if (token) {
      headers['Authorization'] = `Bearer ${token}`;
    }
    
    return headers;
  }

  async saveCode(request: SaveCodeRequest): Promise<SaveResponse> {
    try {
      const response = await fetch(`${this.baseUrl}/save/code`, {
        method: 'POST',
        headers: this.getAuthHeaders(),
        body: JSON.stringify(request)
      });

      if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`);
      }

      return await response.json();
    } catch (error) {
      console.error('Failed to save code:', error);
      throw error;
    }
  }

  async saveLogs(request: SaveLogRequest): Promise<SaveResponse> {
    try {
      const response = await fetch(`${this.baseUrl}/save/logs`, {
        method: 'POST',
        headers: this.getAuthHeaders(),
        body: JSON.stringify(request)
      });

      if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`);
      }

      return await response.json();
    } catch (error) {
      console.error('Failed to save logs:', error);
      throw error;
    }
  }

  async saveSession(request: SaveSessionRequest): Promise<SaveResponse> {
    try {
      const response = await fetch(`${this.baseUrl}/save/session`, {
        method: 'POST',
        headers: this.getAuthHeaders(),
        body: JSON.stringify(request)
      });

      if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`);
      }

      return await response.json();
    } catch (error) {
      console.error('Failed to save session:', error);
      throw error;
    }
  }

  async getSavedCodes(page: number = 1, perPage: number = 20, search?: string, language?: string): Promise<SavedItemsResponse> {
    try {
      const params = new URLSearchParams({
        page: page.toString(),
        per_page: perPage.toString()
      });

      if (search) params.append('search', search);
      if (language) params.append('language', language);

      const response = await fetch(`${this.baseUrl}/saved/codes?${params}`, {
        headers: this.getAuthHeaders()
      });

      if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`);
      }

      return await response.json();
    } catch (error) {
      console.error('Failed to get saved codes:', error);
      throw error;
    }
  }

  async getSavedLogs(
    page: number = 1, 
    perPage: number = 20, 
    search?: string, 
    logLevel?: string, 
    source?: string
  ): Promise<SavedItemsResponse> {
    try {
      const params = new URLSearchParams({
        page: page.toString(),
        per_page: perPage.toString()
      });

      if (search) params.append('search', search);
      if (logLevel) params.append('log_level', logLevel);
      if (source) params.append('source', source);

      const response = await fetch(`${this.baseUrl}/saved/logs?${params}`, {
        headers: this.getAuthHeaders()
      });

      if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`);
      }

      return await response.json();
    } catch (error) {
      console.error('Failed to get saved logs:', error);
      throw error;
    }
  }

  async deleteSavedCode(codeId: string): Promise<{ success: boolean; message: string }> {
    try {
      const response = await fetch(`${this.baseUrl}/saved/code/${codeId}`, {
        method: 'DELETE',
        headers: this.getAuthHeaders()
      });

      if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`);
      }

      return await response.json();
    } catch (error) {
      console.error('Failed to delete saved code:', error);
      throw error;
    }
  }

  async deleteSavedLog(logId: string): Promise<{ success: boolean; message: string }> {
    try {
      const response = await fetch(`${this.baseUrl}/saved/log/${logId}`, {
        method: 'DELETE',
        headers: this.getAuthHeaders()
      });

      if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`);
      }

      return await response.json();
    } catch (error) {
      console.error('Failed to delete saved log:', error);
      throw error;
    }
  }
}

const saveService = new SaveService();
export default saveService;

export type {
  SaveCodeRequest,
  SaveLogRequest,
  SaveSessionRequest,
  SaveResponse,
  SavedCode,
  SavedLog,
  SavedItem,
  SavedItemsResponse
};